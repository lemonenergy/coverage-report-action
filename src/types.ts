type CoverageJsonProperties = {
  statements: {
    covered: number
    total: number
    pct: number
  }
  branches: {
    covered: number
    total: number
    pct: number
  }
  functions: {
    covered: number
    total: number
    pct: number
  }
  lines: {
    covered: number
    total: number
    pct: number
  }
}

export type CoverageJson = {
  total: CoverageJsonProperties
  [file: string]: CoverageJsonProperties
}
