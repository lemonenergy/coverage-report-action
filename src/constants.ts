export const REPORT_TITLE = 'ðŸ“Š Test Coverage Report'
