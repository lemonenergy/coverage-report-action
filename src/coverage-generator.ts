import * as core from '@actions/core'
import { mkdirSync, readFileSync, writeFileSync } from 'fs'
import { resolve } from 'path'
import { REPORT_TITLE } from './constants'
import { getChangedFiles } from './github.utils'
import type { CoverageJson } from './types'

const footNote = `> [!NOTE]
> This report was generated by [Coverage Report Action](https://github.com/lemonenergy/coverage-report-action).`

// This guarantees that a PR with only test files will still have coverage data.
// It also increase range of included files when there is a snapshot.
// Only works when tests/snapshots are on the same directory.
const simplifyPath = (path: string) => {
  // Mock for testing purposes.
  const testPath = process.env.VITEST ? '/home/runner/work/' : null

  let transformedPath = path
    .replace(testPath ?? process.cwd(), '') // CI path
    .replace(/^\.\//, '') // Relative path

  // Target file for coverage from snapshots paths.
  //    i.e.: src/Example/__snapshots__/index.test.tsx.snap > src/Example
  if (transformedPath.endsWith('.snap'))
    transformedPath = transformedPath.replace(/\/__snapshots__\/.*$/, '') // Remove everything after `__snapshots__/`

  // Target file for coverage from test files paths.
  //    i.e.: src/Example/index.test.tsx > src/Example/index.tsx
  if (transformedPath.includes('.test.'))
    transformedPath = transformedPath.replace(/\.test\./, '.') // Replace `.test.` with `.`

  return transformedPath
}

const getJsonCoverage = (path: string) => {
  try {
    const workspacePath = process.env.GITHUB_WORKSPACE || process.cwd()
    const resolvedPath = resolve(workspacePath, path)

    core.info(`Looking for coverage file at: ${resolvedPath}`)

    const coverage = JSON.parse(readFileSync(resolvedPath, 'utf8'))

    if (!coverage.total) {
      core.error('No total coverage data found.')
      return null
    }

    return coverage as CoverageJson
  } catch (error) {
    core.error(error instanceof Error ? error.message : String(error), {
      title: 'Error reading coverage file:'
    })
    return null
  }
}

const makeToggleMarkdown = (
  title: string,
  content: string
) => `<details><summary><strong>${title}</strong></summary>

<br />

${content}

</details>\n\n`

const formatCoverage = (pct: number) => {
  if (pct >= 90) return `${pct}% ðŸ”µ` // Green for >= 95%
  if (pct >= 85) return `${pct}% ðŸŸ¢` // Blue for >= 85%
  if (pct >= 70) return `${pct}% ðŸŸ¡` // Yellow for >= 75%
  if (pct >= 50) return `${pct}% ðŸŸ ` // Orange for >= 50%
  return `${pct}% ðŸ”´` // Red for < 50%
}

const makeCoverageMarkdown = (coverage: CoverageJson, paths?: string[]) => {
  let report = `## ${REPORT_TITLE}\n\n`

  report += '### Overall coverage\n\n'

  const { total, ...files } = coverage

  report += '| Metric     | Covered/Total | Status |\n'
  report += '| ---------- | ---------- | ---------- |\n'
  report += `| Statements  | ${total.statements.covered}/${total.statements.total} | ${formatCoverage(total.statements.pct)} |\n`
  report += `| Branches    | ${total.branches.covered}/${total.branches.total} | ${formatCoverage(total.branches.pct)} |\n`
  report += `| Functions   | ${total.functions.covered}/${total.functions.total} | ${formatCoverage(total.functions.pct)} |\n`
  report += `| Lines       | ${total.lines.covered}/${total.lines.total} | ${formatCoverage(total.lines.pct)} |\n\n`

  const filesCoverageLinesData = Object.entries(files)
    .filter(([file]) => !!paths?.find((it) => file.includes(simplifyPath(it))))
    .map(([file, data]) => ({
      ...data,
      path: file
    }))

  report += '### Diff coverage report\n\n'

  if (filesCoverageLinesData.length <= 0) {
    report += 'No files found with coverage data.\n\n'
    report += footNote
    return report
  }

  let coverageRender: string = ''

  coverageRender += '| File | Lines | Coverage |\n'
  coverageRender += '| ----- | ----- | ----- |\n'

  filesCoverageLinesData.forEach((file) => {
    coverageRender += `| ${simplifyPath(file.path)} | ${file.lines.covered}/${file.lines.total} | ${formatCoverage(file.lines.pct)} |\n`
  })

  if (filesCoverageLinesData.length > 10) {
    core.info(
      'More than 10 files with coverage data found. Report will be done inside a toggle.'
    )
    report += makeToggleMarkdown('Coverage by file', coverageRender)
  } else {
    coverageRender += '\n'
    report += coverageRender
  }

  report += footNote

  return report
}

export const generateCoverageReport = async (path: string, token?: string) => {
  const jsonCoverage = getJsonCoverage(path)

  if (!jsonCoverage)
    throw new Error(`No coverage json data found on path ${path}`)

  const diffFilesPaths = await getChangedFiles(token)

  const report = makeCoverageMarkdown(jsonCoverage, diffFilesPaths)

  core.info('Coverage report generated successfully.')

  const dirPath = './coverage-report-action-output'
  const filePath = `${dirPath}/coverage-report.md`

  mkdirSync(dirPath, { recursive: true })
  writeFileSync(filePath, report, 'utf8')

  return {
    filePath,
    report
  }
}
