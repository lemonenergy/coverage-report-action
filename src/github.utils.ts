import * as core from '@actions/core'
import { context, getOctokit } from '@actions/github'
import { readFileSync } from 'fs'
import { REPORT_TITLE } from './constants'

export const getChangedFiles = async (token?: string): Promise<string[]> => {
  if (!token) throw new Error('GitHub token is required')

  const octokit = getOctokit(token)

  const pullRequest = context.payload.pull_request

  if (!pullRequest)
    throw new Error('This action should only run on pull requests')

  const owner = context.repo.owner
  const repo = context.repo.repo
  const prNumber = pullRequest.number

  try {
    core.info(`Getting changed files for PR #${prNumber}`)

    // Get the list of files changed in the PR
    const { data: files } = await octokit.rest.pulls.listFiles({
      owner,
      per_page: 100,
      pull_number: prNumber,
      repo
    })

    const changedFilePaths = files.map((file) => file.filename)

    core.info(`Found ${changedFilePaths.length} changed files`)
    core.debug(`Changed files: ${changedFilePaths.join(', ')}`)

    return changedFilePaths
  } catch (error) {
    core.setFailed(
      `Failed to get changed files: ${error instanceof Error ? error.message : String(error)}`
    )
    return []
  }
}

export const updateCoverageComment = async (
  coverageMdPath: string,
  token?: string
) => {
  if (!token) throw new Error('GitHub token is required')

  const octokit = getOctokit(token)

  const pullRequest = context.payload.pull_request

  if (!pullRequest)
    throw new Error('This action should only run on pull requests')

  const owner = context.repo.owner
  const repo = context.repo.repo
  const prNumber = pullRequest.number

  const coverageReport = readFileSync(coverageMdPath, 'utf-8')

  try {
    // Check for existing comments by the bot
    const { data: comments } = await octokit.rest.issues.listComments({
      issue_number: prNumber,
      owner,
      repo
    })

    const coverageBotComment = comments.find(
      (comment) =>
        comment.user?.login === 'github-actions[bot]' &&
        comment.body?.includes(REPORT_TITLE)
    )

    if (coverageBotComment) {
      // Remove the existing comment to replace with a new one
      core.info(
        `Removing existing coverage comment with ID ${coverageBotComment.id}`
      )
      await octokit.rest.issues.deleteComment({
        comment_id: coverageBotComment.id,
        owner,
        repo
      })
    }

    await octokit.rest.issues.createComment({
      body: coverageReport,
      issue_number: prNumber,
      owner,
      repo
    })

    core.info('Coverage report comment upserted successfully.')
  } catch (error) {
    core.setFailed(
      `Failed to upsert coverage comment: ${error instanceof Error ? error.message : String(error)}`
    )
  }
}
