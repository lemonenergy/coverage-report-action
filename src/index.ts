import { getInput, info, setFailed } from '@actions/core'
import { exec } from '@actions/exec'
import { context } from '@actions/github'
import { generateCoverageReport } from './coverage-generator'
import { updateCoverageComment } from './github.utils'

const run = async () => {
  try {
    info('Release action started...')

    const githubToken = getInput('github-token')

    await exec('git', [
      'config',
      '--global',
      'user.email',
      '"github-actions[bot]@users.noreply.github.com"'
    ])
    await exec('git', [
      'config',
      '--global',
      'user.name',
      '"github-actions[bot]"'
    ])

    await exec('git', [
      'remote',
      'set-url',
      'origin',
      `https://x-access-token:${githubToken}@github.com/${context.repo.owner}/${context.repo.repo}.git`
    ])

    info('Git config completed...')

    const coverageJsonPath = getInput('coverage-json-path')

    const { filePath } = await generateCoverageReport(
      coverageJsonPath,
      githubToken
    )

    await updateCoverageComment(filePath, githubToken)

    info('Coverage report generated and comment updated successfully.')
  } catch (error) {
    setFailed(
      error instanceof Error ? error.message : `Unknown error: ${error}`
    )
  }
}

run()
