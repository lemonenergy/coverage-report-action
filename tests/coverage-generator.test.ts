import { rmSync } from 'fs'
import { afterAll, describe, expect, it, vi } from 'vitest'
import { generateCoverageReport } from '../src/coverage-generator'
import { getChangedFiles } from '../src/github.utils'

const dummyCoveragePath = 'tests/fixtures/coverage-summary.json'

vi.mock('../src/github.utils', () => ({
  getChangedFiles: vi.fn(() => Promise.resolve([]))
}))

vi.mock('@actions/core')

describe('generateCoverageReport()', () => {
  it('should throw if the coverage json is not found in path', async () => {
    const fakePath = 'fakepath/coverage-summary.json'
    const fn = generateCoverageReport(fakePath, '')

    await expect(fn).rejects.toThrow(
      `No coverage json data found on path ${fakePath}`
    )
  })

  it('should generate the coverage successfully when there is no coverage for diff', async () => {
    const { filePath, report } = await generateCoverageReport(
      dummyCoveragePath,
      ''
    )

    expect(filePath).toBe('./coverage-report-action-output/coverage-report.md')
    expect(report).toContain('No files found with coverage data.')
    expect(report).toMatchSnapshot()
  })

  it('should generate the coverage successfully when there is coverage for diff', async () => {
    vi.mocked(getChangedFiles).mockResolvedValueOnce([
      'src/index.ts',
      'src/green.ts',
      'src/yellow.ts',
      'src/orange.ts',
      'src/red.ts'
    ])

    const { filePath, report } = await generateCoverageReport(
      dummyCoveragePath,
      ''
    )

    expect(filePath).toBe('./coverage-report-action-output/coverage-report.md')
    expect(report).toContain('| src/index.ts | 10/10 | 100% 🔵 |')
    expect(report).toContain('| src/green.ts | 88/100 | 88% 🟢 |')
    expect(report).toContain('| src/yellow.ts | 77/100 | 77% 🟡 |')
    expect(report).toContain('| src/orange.ts | 58/100 | 58% 🟠 |')
    expect(report).toContain('| src/red.ts | 45/100 | 45% 🔴 |')
    expect(report).toMatchSnapshot()
  })

  it('should generate the coverage successfully when there is coverage for diff and more than 10 files', async () => {
    vi.mocked(getChangedFiles).mockResolvedValueOnce([
      'src/index.ts',
      'src/green.ts',
      'src/yellow.ts',
      'src/orange.ts',
      'src/red.ts',
      'src/dummy1.ts',
      'src/dummy2.ts',
      'src/dummy3.ts',
      'src/dummy4.ts',
      'src/dummy5.ts',
      'src/dummy6.ts'
    ])

    const { filePath, report } = await generateCoverageReport(
      dummyCoveragePath,
      ''
    )

    expect(filePath).toBe('./coverage-report-action-output/coverage-report.md')
    expect(report)
      .toContain(`<details><summary><strong>Coverage by file</strong></summary>

<br />

| File | Lines | Coverage Status |
| ----- | ----- | ----- |
| src/index.ts | 10/10 | 100% 🔵 |
| src/green.ts | 88/100 | 88% 🟢 |
| src/yellow.ts | 77/100 | 77% 🟡 |
| src/orange.ts | 58/100 | 58% 🟠 |
| src/red.ts | 45/100 | 45% 🔴 |
| src/dummy1.ts | 1/1 | 100% 🔵 |
| src/dummy2.ts | 1/1 | 100% 🔵 |
| src/dummy3.ts | 1/1 | 100% 🔵 |
| src/dummy4.ts | 1/1 | 100% 🔵 |
| src/dummy5.ts | 1/1 | 100% 🔵 |
| src/dummy6.ts | 1/1 | 100% 🔵 |


</details>`)
    expect(report).toMatchSnapshot()
  })

  afterAll(() => {
    // Remove the local generated `coverage-report-action-output` directory
    rmSync('./coverage-report-action-output', { force: true, recursive: true })
  })
})
