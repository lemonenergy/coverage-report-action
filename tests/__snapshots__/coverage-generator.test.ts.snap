// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateCoverageReport() > should generate the coverage successfully when there is coverage for diff 1`] = `
"## 📊 Test Coverage Report

### Overall coverage

| Metric     | Covered/Total | Status |
| ---------- | ---------- | ---------- |
| Statements  | 5/5 | 100% 🔵 |
| Branches    | 5/7 | 71.42% 🟡 |
| Functions   | 5/5 | 100% 🔵 |
| Lines       | 379/410 | 92.43% 🔵 |

### Diff coverage report

| File | Lines | Coverage |
| ----- | ----- | ----- |
| src/index.ts | 10/10 | 100% 🔵 |
| src/green.ts | 88/100 | 88% 🟢 |
| src/yellow.ts | 77/100 | 77% 🟡 |
| src/orange.ts | 58/100 | 58% 🟠 |
| src/red.ts | 45/100 | 45% 🔴 |

> [!NOTE]
> This report was generated by [Coverage Report Action](https://github.com/lemonenergy/coverage-report-action)."
`;

exports[`generateCoverageReport() > should generate the coverage successfully when there is coverage for diff and more than 10 files 1`] = `
"## 📊 Test Coverage Report

### Overall coverage

| Metric     | Covered/Total | Status |
| ---------- | ---------- | ---------- |
| Statements  | 5/5 | 100% 🔵 |
| Branches    | 5/7 | 71.42% 🟡 |
| Functions   | 5/5 | 100% 🔵 |
| Lines       | 379/410 | 92.43% 🔵 |

### Diff coverage report

<details><summary><strong>Coverage by file</strong></summary>

<br />

| File | Lines | Coverage |
| ----- | ----- | ----- |
| src/index.ts | 10/10 | 100% 🔵 |
| src/green.ts | 88/100 | 88% 🟢 |
| src/yellow.ts | 77/100 | 77% 🟡 |
| src/orange.ts | 58/100 | 58% 🟠 |
| src/red.ts | 45/100 | 45% 🔴 |
| src/dummy1.ts | 1/1 | 100% 🔵 |
| src/dummy2.ts | 1/1 | 100% 🔵 |
| src/dummy3.ts | 1/1 | 100% 🔵 |
| src/dummy4.ts | 1/1 | 100% 🔵 |
| src/dummy5.ts | 1/1 | 100% 🔵 |
| src/dummy6.ts | 1/1 | 100% 🔵 |


</details>

> [!NOTE]
> This report was generated by [Coverage Report Action](https://github.com/lemonenergy/coverage-report-action)."
`;

exports[`generateCoverageReport() > should generate the coverage successfully when there is no coverage for diff 1`] = `
"## 📊 Test Coverage Report

### Overall coverage

| Metric     | Covered/Total | Status |
| ---------- | ---------- | ---------- |
| Statements  | 5/5 | 100% 🔵 |
| Branches    | 5/7 | 71.42% 🟡 |
| Functions   | 5/5 | 100% 🔵 |
| Lines       | 379/410 | 92.43% 🔵 |

### Diff coverage report

No files found with coverage data.

> [!NOTE]
> This report was generated by [Coverage Report Action](https://github.com/lemonenergy/coverage-report-action)."
`;
